File: app/src/main/java/aleksandrkim/ArchComponentsTest/Db/NoteDao.java
package aleksandrkim.ArchComponentsTest.Db;

import android.arch.paging.DataSource;
import android.arch.persistence.room.Dao;
import android.arch.persistence.room.Insert;
import android.arch.persistence.room.Query;
import android.arch.persistence.room.Transaction;
import android.arch.persistence.room.Update;

import static android.arch.persistence.room.OnConflictStrategy.IGNORE;

/**
 * Created by Aleksandr Kim on 11 Apr, 2018 6:41 PM for ArchComponentsTest
 */

@Dao
public abstract class NoteDao {

    @Query("SELECT * FROM Note ORDER BY lastModified DESC")
    public abstract DataSource.Factory<Integer, Note> getNotesPagedLastModifiedFirst();

    @Query("SELECT * FROM Note ORDER BY createdTime DESC")
    public abstract DataSource.Factory<Integer, Note> getNotesPagedLastCreatedFirst();

    @Query("SELECT * FROM Note WHERE id = :id")
    public abstract Note getNoteById(int id);

    @Insert (onConflict = IGNORE)
    public abstract void add(Note note);

    @Update (onConflict = IGNORE)
    public abstract void update(Note note);

    @Transaction
    public void upsert(Note note) {
        add(note);
        update(note);
    }

    @Query("DELETE FROM Note")
    public abstract void deleteAll();

    @Query("DELETE FROM Note WHERE id = :id")
    public abstract void delete(int id);


}

File: app/src/main/java/aleksandrkim/MinimalNotes/Db/NoteDao.kt
package aleksandrkim.MinimalNotes.Db

import android.arch.paging.DataSource
import android.arch.persistence.room.Dao
import android.arch.persistence.room.Query

/**
 * Created by Aleksandr Kim on 11 Apr, 2018 6:41 PM for ArchComponentsTest
 */

@Dao
abstract class NoteDao : BaseDao<Note>() {

    @Query("SELECT * FROM " + Note.TABLE_NAME + " ORDER BY createdTime DESC")
    abstract fun getNotesPagedLastCreatedFirst(): DataSource.Factory<Int, Note>

    @Query("SELECT * FROM " + Note.TABLE_NAME + " WHERE id = :id")
    abstract fun getNoteById(id: Int): Note?

    @Query("SELECT * FROM " + Note.TABLE_NAME + " ORDER BY createdTime DESC LIMIT 1")
    abstract fun getLatestNote(): Note

    @Query("DELETE FROM " + Note.TABLE_NAME)
    abstract fun deleteAll()

    @Query("DELETE FROM " + Note.TABLE_NAME + " WHERE id = :id")
    abstract fun delete(id: Int)

    @Query("SELECT COUNT(*) FROM " + Note.TABLE_NAME)
    abstract override fun count(): Int

}
